backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2687,24 +2764,42 @@ public Response commitArtifactPart(CommitArtifactPart request) throws ModelDBExc
       throws ModelDBException {
     List<PartETag> partETags;
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
-      ArtifactEntity artifactEntity = getArtifactEntity(session, request.getId(), request.getKey());
-      if (artifactEntity.getUploadId() == null) {
-        String message = "Multipart wasn't initialized";
-        LOGGER.info(message);
-        throw new ModelDBException(message, io.grpc.Status.Code.FAILED_PRECONDITION);
+      try (AutoCloseable ignored =
+          acquireWriteLock(buildArtifactLockKey(request.getId(), request.getKey()))) {
+        ArtifactEntity artifactEntity =
+            getArtifactEntity(session, request.getId(), request.getKey());
+        if (artifactEntity.getUploadId() == null) {
+          String message = "Multipart wasn't initialized OR Multipart artifact already committed";
+          LOGGER.info(message);
+          throw new ModelDBException(message, Code.FAILED_PRECONDITION);
+        }
+        Set<ArtifactPartEntity> artifactPartEntities =
+            VersioningUtils.getArtifactPartEntities(
+                session,
+                String.valueOf(artifactEntity.getId()),
+                ArtifactPartEntity.EXP_RUN_ARTIFACT);
+        partETags =
+            artifactPartEntities.stream()
+                .map(ArtifactPartEntity::toPartETag)
+                .collect(Collectors.toList());
+        commitMultipartFunction.apply(
+            artifactEntity.getPath(), artifactEntity.getUploadId(), partETags);
+        session.beginTransaction();
+        artifactEntity.setUploadCompleted(true);
+        artifactEntity.setUploadId(null);
+        artifactPartEntities.forEach(session::delete);
+        artifactPartEntities.clear();
```

Comment:
There was a change needed for python client to keep artfact parts after upload completion: https://vertaai.atlassian.net/browse/VR-4760 to remove these lines from here.

Are you sure that these lines are necessary? There is a feature on client to continue cancelled upload. I think these lines may break it.

Reply:
Not necessary, but I see stale data in the artifact part table so I thought to do a cleanup but if the client has a feature to continue canceled upload then I can revert this change.